### ISSUE TITLE
Excuse me, how can I solve the problem that the confidence level is only 0.1 after switching to the ONNX model?

### ISSUE THREAD

**Issue by CanhaoL** [ultralytics/ultralytics#18896]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

![Image](https://github.com/user-attachments/assets/ee89874f-1b06-4319-80d0-9f6bc96c903a)

### Environment

[2025/01/26 15:52:38] ppocr DEBUG: Namespace(alpha=1.0, alphacolor=(255, 255, 255), benchmark=False, beta=1.0, binarize=False, cls_batch_num=6, cls_image_shape='3, 48, 192', cls_model_dir='./weights/ocr/ch_ppocr_mobile_v2.0_cls_infer/', cls_thresh=0.9, cpu_threads=10, crop_res_save_dir='./output', det=True, det_algorithm='DB', det_box_type='quad', det_db_box_thresh=0.6, det_db_score_mode='fast', det_db_thresh=0.3, det_db_unclip_ratio=1.5, det_east_cover_thresh=0.1, det_east_nms_thresh=0.2, det_east_score_thresh=0.8, det_limit_side_len=960, det_limit_type='max', det_model_dir='/home/tony/.paddleocr/whl/det/en/en_PP-OCRv3_det_infer', det_pse_box_thresh=0.85, det_pse_min_area=16, det_pse_scale=1, det_pse_thresh=0, det_sast_nms_thresh=0.2, det_sast_score_thresh=0.5, draw_img_save_dir='./inference_results', drop_score=0.5, e2e_algorithm='PGNet', e2e_char_dict_path='./ppocr/utils/ic15_dict.txt', e2e_limit_side_len=768, e2e_limit_type='max', e2e_model_dir=None, e2e_pgnet_mode='fast', e2e_pgnet_score_thresh=0.5, e2e_pgnet_valid_set='totaltext', enable_mkldnn=False, formula=False, formula_algorithm='LaTeXOCR', formula_batch_num=1, formula_char_dict_path=None, formula_model_dir=None, fourier_degree=5, gpu_id=0, gpu_mem=500, help='==SUPPRESS==', image_dir=None, image_orientation=False, invert=False, ir_optim=True, kie_algorithm='LayoutXLM', label_list=['0', '180'], lang='en', layout=True, layout_dict_path=None, layout_model_dir=None, layout_nms_threshold=0.5, layout_score_threshold=0.5, max_batch_size=10, max_text_length=25, merge_no_span_structure=True, min_subgraph_size=15, mode='structure', ocr=True, ocr_order_method=None, ocr_version='PP-OCRv4', output='./output', page_num=0, precision='fp32', process_id=0, re_model_dir=None, rec=True, rec_algorithm='SVTR_LCNet', rec_batch_num=6, rec_char_dict_path='./weights/ocr/ppocr_keys_v1_fhhx.txt', rec_image_inverse=True, rec_image_shape='3, 48, 320', rec_model_dir='./weights/ocr/0510/', recovery=False, recovery_to_markdown=False, return_word_box=False, save_crop_res=False, save_log_path='./log_output/', savefile=False, scales=[8, 16, 32], ser_dict_path='../train_data/XFUND/class_list_xfun.txt', ser_model_dir=None, show_log=True, sr_batch_num=1, sr_image_shape='3, 32, 128', sr_model_dir=None, structure_version='PP-StructureV2', table=True, table_algorithm='TableAttn', table_char_dict_path=None, table_max_len=488, table_model_dir=None, total_process_num=1, type='ocr', use_angle_cls=False, use_dilation=False, use_gpu=True, use_mlu=False, use_mp=False, use_npu=False, use_onnx=False, use_pdf2docx_api=False, use_pdserving=False, use_space_char=True, use_tensorrt=False, use_visual_backbone=True, use_xpu=False, vis_font_path='./doc/fonts/simfang.ttf', warmup=False)
[2025/01/26 15:52:38] ppocr WARNING: The first GPU is used for inference by default, GPU ID: 0
[2025/01/26 15:52:39] ppocr WARNING: The first GPU is used for inference by default, GPU ID: 0
Ultralytics 8.3.66 🚀 Python-3.8.8 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 3090, 24132MiB)

### Minimal Reproducible Example
```python
import cv2
import math
import copy
import torch
import time
import os
import onnxruntime as ort
from paddleocr import PaddleOCR
import concurrent.futures


# 将 YOLO 模型转换为 ONNX 模型
def export_to_onnx(weights):
    from ultralytics import YOLO
    model = YOLO(weights)
    try:
        model.export(format='onnx')
        print("ONNX 模型转换成功。")
    except Exception as e:
        print(f"ONNX 模型转换失败: {e}")


class YOLO_det:
    def __init__(self, weights, imgsz=640, conf_thres=0.1, iou_thres=0.25, max_det=1000):  # 提高置信度阈值
        if torch.cuda.is_available() and torch.cuda.device_count() > 0:
            providers = [
                ('CUDAExecutionProvider', {
                    'device_id': 0,
                    'arena_extend_strategy': 'kNextPowerOfTwo',
                    'gpu_mem_limit': 4 * 1024 * 1024 * 1024,  # 4GB 内存限制
                    'cudnn_conv_algo_search': 'EXHAUSTIVE',
                    'do_copy_in_default_stream': True,
                })
            ]
        else:
            providers = ['CPUExecutionProvider']

        onnx_weights = weights.replace('.pt', '.onnx')
        if not os.path.exists(onnx_weights):
            export_to_onnx(weights)

        self.session = ort.InferenceSession(onnx_weights, providers=providers)
        self.imgsz = imgsz
        self.conf = conf_thres
        self.iou = iou_thres
        self.max_det = max_det

    def preprocess(self, img):
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        img = cv2.resize(img, (self.imgsz, self.imgsz))
        img = img.transpose(2, 0, 1)
        img = img[None]
        img = img.astype('float32') / 255.0
        return img

    def detect(self, img):
        input_name = self.session.get_inputs()[0].name
        input_img = self.preprocess(img)
        try:
            outputs = self.session.run(None, {input_name: input_img})
            print(f"推理输出形状: {[o.shape for o in outputs]}")  # 打印输出形状
            print(f"推理输出部分内容: {outputs[0][0, :5, :]}")  # 打印部分输出内容
        except Exception as e:
            print(f"推理过程中出现异常: {e}")
            return []

        # 假设输出只有一个数组，需要根据实际情况解析
        output = outputs[0]
        boxes = []
        confidences = []
        # 根据实际输出格式调整解析逻辑
        if output.ndim == 3:
            num_detections = output.shape[1]
            for i in range(num_detections):
                # 假设前 4 列是边界框信息，第 5 列是置信度
                box = output[0, i, :4]
                conf = output[0, i, 4]
                boxes.append(box)
                confidences.append(conf)
        else:
            print(f"不支持的输出形状: {output.shape}")
            return []

        return_list = []
        for box, conf in zip(boxes, confidences):
            if conf > self.conf:
                xyxy = box
                x1 = math.ceil(xyxy[0])
                y1 = math.ceil(xyxy[1])
                x2 = math.ceil(xyxy[2])
                y2 = math.ceil(xyxy[1])
                x3 = math.ceil(xyxy[2])
                y3 = math.ceil(xyxy[3])
                x4 = math.ceil(xyxy[0])
                y4 = math.ceil(xyxy[3])
                return_list.append([[x1, y1], [x2, y2], [x3, y3], [x4, y4]])
        if not return_list:
            print("未检测到目标。")
        return return_list


def sorted_boxes(dt_boxes):
    sorted_boxes = sorted(dt_boxes, key=lambda x: (x[0][1], x[0][0]))
    _boxes = [[[int(num) for num in sub_list] for sub_list in main_list]
              for main_list in sorted_boxes]
    for i in range(len(dt_boxes) - 1):
        for j in range(i, -1, -1):
            if abs(_boxes[j + 1][0][1] - _boxes[j][0][1]) < 10 and (
                    _boxes[j + 1][0][0] < _boxes[j][0][0]
            ):
                tmp = _boxes[j]
                _boxes[j] = _boxes[j + 1]
                _boxes[j + 1] = tmp
            else:
                break
    return _boxes


def _4point2xyxy(points):
    list_out_xyxy = []
    for point in points:
        x_coords, y_coords = zip(*point)
        min_x, max_x = min(x_coords), max(x_coords)
        min_y, max_y = min(y_coords), max(y_coords)
        rectangle = [int(min_x), int(min_y), int(max_x), int(max_y)]
        list_out_xyxy.append(rectangle)
    return list_out_xyxy


def process_image(index, image_path, ocr_rec, yolo_det):
    start = time.time()  # 记录开始时间
    try:
        det_img = cv2.imread(image_path)
        if det_img is None:
            print(f"无法读取图片: {image_path}")
            return
    except Exception as e:
        print(f"读取图片 {image_path} 时出现错误: {e}")
        return
    out_yolo_det = yolo_det.detect(det_img)
    if not out_yolo_det:
        print(f"在图片 {image_path} 中未检测到目标。")
    out_yolo_det = sorted_boxes(out_yolo_det)
    list_ocr_det_bbox_xyxy = _4point2xyxy(out_yolo_det)
    show_img = copy.deepcopy(det_img)

    for i in range(len(list_ocr_det_bbox_xyxy)):
        # xyxy 坐标
        x1, y1, x2, y2 = list_ocr_det_bbox_xyxy[i]

        # 检查截取区域是否有效
        if x2 > x1 and y2 > y1:
            # 截取文本小图
            ocr_rec_det_img = det_img[y1:y2, x1:x2]

            one_ocr_rec_out = ocr_rec.ocr(ocr_rec_det_img, det=False, cls=False)
            print(one_ocr_rec_out)

            # 绘制 bbox
            show_img = cv2.rectangle(show_img, (x1, y1), (x2, y2), (0, 0, 255), 1)

            # 设置字体、大小、颜色和线条粗细
            font = cv2.FONT_HERSHEY_SIMPLEX

            # 绘制文本
            show_img = cv2.putText(show_img, str(i), (x1, y1 + 20), font, 0.8, (0, 255, 0), 2)

    # 确保 output 文件夹存在
    output_folder = 'output_onnx'
    if not os.path.exists(output_folder):
        try:
            os.makedirs(output_folder)
            print(f"成功创建输出文件夹: {output_folder}")
        except OSError as e:
            print(f"创建输出文件夹时出错: {e}")
            return

    # 保存图片
    output_path = os.path.join(output_folder, f'out_image{index + 1}.jpg')
    if cv2.imwrite(output_path, show_img):
        print(f"图片已成功保存到: {output_path}")
    else:
        print(f"无法保存图片到: {output_path}，请检查文件权限或路径是否正确。")

    end = time.time()  # 记录结束时间
    elapsed = end - start  # 计算该图片处理耗时
    print(f"图片 {image_path} 处理耗时: {elapsed:.2f} 秒")
    print("\n", "=" * 200, "\n")


if __name__ == "__main__":
    start_time = time.time()

    # 文本识别_权重
    rec_model_dir = './weights/ocr/0510/'

    # 文本字典
    rec_char_dict_path = './weights/ocr/ppocr_keys_v1_fhhx.txt'

    # 方向分类器
    cls_model_dir = './weights/ocr/ch_ppocr_mobile_v2.0_cls_infer/'

    # yolo_det 权重目录
    weighs = './weights/best.pt'

    # 加载 文本检测权重...
    ocr_rec = PaddleOCR(lang='en', rec_model_dir=rec_model_dir, rec_char_dict_path=rec_char_dict_path,
                        cls_model_dir=cls_model_dir, use_gpu=True)  # 明确指定使用GPU

    # 加载 yolo_det 权重...
    yolo_det = YOLO_det(weighs, imgsz=640)

    # 获取 input_images 文件夹下的所有图片路径
    input_folder = 'input_images'
    if not os.path.exists(input_folder):
        print(f"输入文件夹 {input_folder} 不存在，请检查路径。")
    else:
        image_files = [os.path.join(input_folder, f) for f in os.listdir(input_folder)
                       if f.endswith(('.png', '.jpg', '.jpeg'))]

        if not image_files:
            print(f"在 {input_folder} 中未找到有效的图片文件，请检查文件夹内容。")
        else:
            # 使用线程池并行处理图片，并行数量为 10
            with concurrent.futures.ThreadPoolExecutor(max_workers=10) as executor:
                futures = []
                for index, image_path in enumerate(image_files):
                    future = executor.submit(process_image, index, image_path, ocr_rec, yolo_det)
                    futures.append(future)

                # 等待所有任务完成
                concurrent.futures.wait(futures)

    end_time = time.time()
    elapsed_time = end_time - start_time
    print(f"代码总运行时间: {elapsed_time:.2f} 秒")

    # 一些后续可能添加的收尾操作可以在这里进行
    # 例如，释放一些资源（虽然目前代码里没有明显需要手动释放的资源）
    # 或者做一些数据统计、日志记录等额外工作
    # 下面是一个简单的示例，用于记录本次运行的总时间到一个日志文件中
    log_file_path = "run_log.txt"
    try:
        with open(log_file_path, "a") as log_file:
            log_file.write(f"本次运行于 {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime())} 开始，耗时 {elapsed_time:.2f} 秒。\n")
    except Exception as e:
        print(f"写入日志文件时出现错误: {e}")
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @CanhaoL, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. From your description, you have already provided extensive code snippets and ONNX integration details, which is fantastic 👏. It would also be helpful if you could share the specific ONNX model file and a sample input image (if possible) for us to reproduce the issue on our end.

If this is a custom training ❓ Question or deployment challenge, please ensure you are following best practices for ONNX inference. Some users might experience confidence-related issues when switching to ONNX due to preprocessing mismatches or inference provider configurations. Please confirm that your preprocessing matches the original PyTorch YOLOv8 pipeline, and ensure you're utilizing the latest version of ONNX Runtime and the Ultralytics library.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community and Support

Join the Ultralytics community where it suits you best:
- For real-time chat, have discussions on [Discord](https://discord.com/invite/ultralytics) 🎧.
- Prefer detailed questions and answers? Use [Discourse](https://community.ultralytics.com/).
- Engage with YOLO enthusiasts on our [Subreddit](https://reddit.com/r/Ultralytics).

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will review your issue and provide feedback soon. Thanks for your patience! 😊

**Comment by Y-T-G** [comment]
You can find ONNX example here:
https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-ONNXRuntime
